### ISSUE TITLE
The `sourcemap` option doesn't combine source maps

### ISSUE THREAD

**Issue by mzgol** [olov/ng-annotate#161]
Refs #79.

I have a source file at `app/modules/init.js` that I first pass through babel & save as `app/.tmp/processed-pre-ng-annotate/modules/init.js` with inline source maps & then pass through `ng-annotate` (using `grunt-ng-annotate`) to `app/.tmp/processed/modules/init.js`. Source maps are not combined properly.

The following code combines fine:
```
ng-annotate --add --sourcemap -o app/.tmp/processed/modules/init.js app/.tmp/processed-pre-ng-annotate/modules/init.js
```
The following doesn't:
```js
ngAnnotate(source, {
    add: true,
    remove: false,
    single_quotes: true,
    sourcemap: {
        inline: true,
        inFile: 'app/.tmp/processed-pre-ng-annotate/modules/init.js'
    }
});
```
the map points to `app/.tmp/processed-pre-ng-annotate/modules/init.js`, the `sources` array in the map file points to this file as well. I've looked at what the binary does and it seems it uses a `map` property instead of the `sourcemap` one. I've checked and... the following works fine:
```js
ngAnnotate(source, {
    add: true,
    remove: false,
    single_quotes: true,
    map: {
        inline: true,
        inFile: 'app/.tmp/processed-pre-ng-annotate/modules/init.js'
    }
});
```

This duplication is weird; also, the `map` property is undocumented and the `sourcemap` one has this bug.

Note that since `grunt-ng-annotate` uses the `sourcemap` property, combining maps doesn't work when used through this Grunt plugin.

Tested on `ng-annotate` `0.15.4`.

**Comment by olov** [comment]
Would this issue be resolved if https://github.com/olov/ng-annotate/blob/master/README.md#library-api had `map:` instead of `sourcemap:` in the example? I think that was intended with https://github.com/olov/ng-annotate/commit/26f2974e24a6b869bbc7a1a1a518bf79d31c3dc1

**Comment by mzgol** [comment]
Why keep both properties, though? People are used to many tools that accept an option named `sourcemap` or `sourceMap` so even if only `map` is in the example, I still think many people will use `sourcemap` and will think they did sth wrong as combining doesn't work.

**Comment by mzgol** [comment]
cc @smrq

**Comment by olov** [comment]
I'm leaving this to @smrq but I'm noticing that the API `config.map` and CLI `--sourcemap` are not the same thing. Also - I don't think we can't ditch `config.map` now because some consumers are relying on it (even though the docs were wrong). Not sure what the best solution is but yah this should be cleaned up. BTW thanks for debugging and finding the issue - this will be helpful for lots of ng-annotate users.

**Comment by omsmith** [comment]
https://github.com/olov/ng-annotate/blob/97f29c74925cf3dcaa56356878c756e6c7e21fba/ng-annotate-main.js#L1146

So, it seems to me that `--sourcemap` just doesnt do anything, as I don't see us referencing `config.map` anywhere?

**Comment by olov** [comment]
The `--sourcemap` CLI option works as it should. `ng-annotate-main.js` is the API code and `ng-annotate.js` is the CLI wrapper (which initializes `config.map`)

**Comment by mzgol** [comment]
> Also - I don't think we can't ditch config.map now because some consumers are relying on it (even though the docs were wrong).

If those two worked as aliases that'd be good as well. What's bad is that they work differently meaning that one supports less features.

**Comment by mzgol** [comment]
To make it clear, I think only one of those options, `map` or `sourcemap` should be documented but the other, if left, should be an alias, optionally prinitng a warning advising a developer to change the use to the recommended name.

But for me the most important issue here to have a documented API that works correctly. :) I'll then switch `grunt-ng-annotate` to use it.

**Comment by smrq** [comment]
Yes, I don't really care what the API is as long as it is documented correctly :)  I might have screwed it up, I haven't looked through the history really.  Whatever works :+1:

**Comment by olov** [comment]
Ok, I'll decide then. :) Let's keep the API as-is (`config.map`) and fix the documentation. @mzgol you can update `grunt-ng-annotate` right away without waiting for a new version. (`config.sourcemap` has never worked). I'll close this issue once I've fixed the documentation.

**Comment by mzgol** [comment]
Great, thanks!

**Comment by mzgol** [comment]
@olov It seems that the `map` property works only if the file already contains a map, that is the following:
```js
ngAnnotate(source, {
    add: true,
    remove: false,
    single_quotes: true,
    map: {
        inline: true,
        inFile: 'app/.tmp/processed-pre-ng-annotate/modules/init.js'
    }
});
```
works if and only if the file `app/.tmp/processed-pre-ng-annotate/modules/init.js` was generated from another source file; otherwise no map is generated. On the other hand, the following:
```js
ngAnnotate(source, {
    add: true,
    remove: false,
    single_quotes: true,
    sourcemap: {
        inline: true,
        inFile: 'app/.tmp/processed-pre-ng-annotate/modules/init.js'
    }
});
```
always creates a source map but always a non-combined one.

So there's no API I can use that would work in both scenarios. :P

**Comment by mzgol** [comment]
I added more detailed source maps tests to `grunt-ng-annotate` & changed `.sourcemap` to `.map` on a [combine-sourcemaps branch](https://github.com/mzgol/grunt-ng-annotate/commit/cbb45cfdeedc3d20abb056326d9e5359a96d4b39); see the result:
https://travis-ci.org/mzgol/grunt-ng-annotate/jobs/62225293
If I change `.map` back to `.sourcemap`, the last test starts passing but other 2 sourcemap tests fail.

**Comment by olov** [comment]
Ok. I haven't taken part in the sourcemap implementation of ng-annotate (other than nagging on perf) so I'd prefer a PR cleaning this up. Takers? =)

**Comment by mzgol** [comment]
I won't have time in the next week or so, especially that I didn't write the code so I'd need to dive in first. If @smrq doesn't find time to fix it, I'll most likely get to it one day.

**Comment by mzgol** [comment]
I'll keep grunt-ng-annotate using the `.sourcemap` part for the time being as I think most people are running ng-annotate on source files and not compiled ones so changing it to `.map` would benefit some people & break a lot more.

**Comment by smrq** [comment]
I'll try and take a look at this in the next day or so.

**Comment by smrq** [comment]
Yeah, that was just a complete fail on my part.  Sorry about that!  PR created to hopefully fix this.  `--sourcemap` and `--sourceroot` are the correct CLI options, and `map: { ... }` is the API.  The docs are updated accordingly.

I tried the update with both incoming and no incoming source maps and it seems to work okay.  Let me know if you have issues.

**Comment by olov** [comment]
Thanks for working on this. Once you're happy with the state of this (and please make sure that the README is updated if not already), I'll merge the PR and will ship out 1.0 to celebrate.

**Comment by mzgol** [comment]
> Once you're happy with the state of this (and please make sure that the README is updated if not already)

@olov It seems the current version of the patch works for both @smrq and me. :) https://github.com/olov/ng-annotate/pull/165#issuecomment-104375379

**Comment by olov** [comment]
Released in 1.0.0. Big thanks!
